### DISCUSSION TITLE
What are some practical/up-to-date  examples for using `dagster-prometheus`?

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#17997] [Q&A] [answered]
Hi there, does anyone have any more practical examples of how to use `dagster-prometheus` to send metrics from Dagster to *Prometheus* for monitoring? <https://docs.dagster.io/_apidocs/libraries/dagster-prometheus>

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1699954362444209?thread_ts=1699954362.444209&cid=C01U954MEER).*

**Comment by alangenfeld** [comment, marked as answer]
`dagster-prometheus` provides a resources that is a very thin wrapper around the the prometheus python client https://github.com/dagster-io/dagster/blob/master/python_modules/libraries/dagster-prometheus/dagster_prometheus/resources.py


You may be able to find examples of using the prometheus python client that you can extrapolate to its use in the context of dagster
https://github.com/prometheus/client_python#exporting-to-a-pushgateway

**Comment by TheEaterr** [comment]
In case anyone finds this useful as I didn't find it documented anywhere and I was struggling with my prometheus metrics sent to the gateway being empty. To actually send anything to the registry one must use the one auto generated in the client resource like so:

```
@op
    def op_send_status_to_prometheus(context, prometheus: PrometheusResource, config: StatusMonitorOpConfig):
        context.log.info("Sending status to Prometheus")
        status = config.status
        run_status_gauge = Gauge('dagster_run_status', 'Status of Dagster runs', ['job_name'], registry=prometheus.registry)        
        run_status_gauge.labels(job_name="run_all_status_monitor").set(1 if status == 'success' else 0)
        
        prometheus.push_to_gateway(job="run_all_status_monitor")
        context.log.info("Status sent to Prometheus")
```

This is probably not the right place to say this but I really think the example included should actually show sending something - anything - and not just a no-op request to really showcase how to use the wrapper.

**Comment by HirofumiTsuda** [comment]
Ran into the same gap (`dagster-prometheus` only pushes to a Pushgateway, so you have to instrument each op yourself, and a crashed/queued run never gets a chance to push at all) and ended up building [dagster-prometheus-exporter](https://github.com/HirofumiTsuda/dagster-prometheus-exporter) instead: a standalone Go binary that polls Dagster's GraphQL API on an interval and exposes `/metrics` for Prometheus to scrape, rather than pushing from inside a run.

The `run_status_gauge` example above is basically `dagster_last_run_info{job_name, location, status}` here — reads the status of each job's most recent completed run straight off Dagster's own state, no manual instrumentation or `push_to_gateway` call needed. `dagster_active_runs` covers in-flight (queued/starting/started) runs the same way.

Also covers schedules, sensors, assets, code locations, and daemon health. Container image and Helm chart published, verified against Dagster 1.13.15.

Linking it here in case it's useful for anyone who hits the same "empty Pushgateway metrics" issue.
